File: src/main/java/net/miarma/mkernel/commands/admin/GmcCommand.java
package net.miarma.mkernel.commands.admin;

import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.GameMode;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.util.MessageUtil;

public class GmcCommand {
    public static void register() {
        CommandWrapper gmcCmd = CommandProvider.getGmcCommand();
        new CommandAPICommand(gmcCmd.getName())
            .withOptionalArguments(
            		CommandProvider.Arguments.playersOptArg().withPermission(
                    gmcCmd.getPermission().others()
                )
            )
            .withShortDescription(gmcCmd.getDescription())
            .withFullDescription(gmcCmd.getDescription())
            .withUsage(gmcCmd.getUsage())
            .executesPlayer((sender,args) -> {
                if(args.rawArgs().length == 0) {
                    sender.setGameMode(GameMode.CREATIVE);
                    MessageUtil.sendMessage(sender, gmcCmd.getMessages()[0], true);
                } else {
                    Bukkit.getPlayer(args.getRaw(0)).setGameMode(GameMode.CREATIVE);
                    MessageUtil.sendMessage(sender, gmcCmd.getMessages()[1], true,
                                                List.of("%player%"), List.of(args.getRaw(0)));
                }
            })
            .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/admin/GmcCommand.kt
package net.miarma.mkernel.command.admin

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.util.PlayerUtil
import org.bukkit.GameMode

@Singleton
class GmcCommand @Inject constructor(
    private val configService: ConfigService,
    private val messageService: MessageService
) {
    fun register() {
        commandAPICommand(configService.getString("commands.gmc.name")) {
            withPermission(configService.getString("commands.gmc.permissions.base"))
            withOptionalArguments(
                PlayerProfileArgument(configService.getString("arguments.player"))
                    .withPermission(configService.getString("commands.gmc.permissions.others"))
            )
            withShortDescription(configService.getString("commands.gmc.description"))
            withUsage(configService.getString("commands.gmc.usage"))
            playerExecutor { sender, args ->
                if (args[0] == null) {
                    sender.gameMode = GameMode.CREATIVE
                    messageService.builder(configService.getString("commands.gmc.messages.self")).withPrefix().send(sender)
                    return@playerExecutor
                }

                val target = PlayerUtil.fromArg(args[0])
                if (target == null || !target.isOnline) {
                    messageService.builder(configService.getString("language.errors.playerNotFound")).withPrefix().send(sender)
                    return@playerExecutor
                }

                target.gameMode = GameMode.CREATIVE
                messageService.builder(configService.getString("commands.gmc.messages.others"))
                    .withPrefix().tag("player", target.name).send(sender)
            }
        }
    }
}